# Data Analysis in Python

In this session we will learn how to properly utilize python's [pandas](https://pandas.pydata.org/) library for data transforming, cleaning, filtering and exploratory data analysis.

## Pandas

Python's Data Analysis Library

Python has long been great for data munging and preparation, but less so for data analysis and modeling. *Pandas* helps fill this gap, enabling you to carry out your entire data analysis workflow in Python.

### Data Structures

Pandas introduces two new data structures to Python: **Series** and **DataFrame**. Both of which are built on top of NumPy.

### Object Creation

A **series**, in *pandas* is a one-dimensional *ndarray* with axis labels (including time series). The axis labels are collectively referred to as the **index**.  
The basic method to create a Series is to call:
```python
s = pd.Series(data, index=index)
```
where *data* is most commonly a dict or a numpy array, and index is a *list* of labels.

In [1]:
from __future__ import print_function
import pandas as pd  # for simplicity we usually refer to pandas as pd
import numpy as np

s = pd.Series([1,3,5,np.nan,6,8])
# By passing a list as the only argument in series, we let pandas create a default integer index
print(s)

0    1.0
1    3.0
2    5.0
3    NaN
4    6.0
5    8.0
dtype: float64


A **DataFrame** is a two-dimensional labeled data structure with columns of potentially different types. It is generally the most commonly used pandas object. You can think of it like a spreadsheet. The easiset way to create a *DataFrame* is to pass in a dictionary of objects.

In [2]:
df = pd.DataFrame({'A' : 1,  # repeats integer for the length of the dataframe
                   'B' : pd.Timestamp('20130102'),  # timestamp datatype, repeats it for the length of the dataframe
                   'C' : pd.Series(1, index=list(range(4)), dtype='float32'),  # creates a series of ones and uses it as a column
                   'D' : np.array([3] * 4,dtype='int32'),  # np.array as a column
                   'E' : pd.Categorical(["test","train","test","train"]),  # categorical data type
                   'F' : 'foo' })  # string, repeats it for the length of the data frame
    
print(df)

   A          B    C  D      E    F
0  1 2013-01-02  1.0  3   test  foo
1  1 2013-01-02  1.0  3  train  foo
2  1 2013-01-02  1.0  3   test  foo
3  1 2013-01-02  1.0  3  train  foo


Once a *DataFrame* object is created, we can add rows to it by the `df.append()` function.

### Viewing Data

In most cases the dataframes are thousands of rows long, we can't view all the data at once.

- We can look at the **first** and **last** entries.

In [3]:
df.head()  # prints first entries (by default 5)

A          B    C  D      E    F
0  1 2013-01-02  1.0  3   test  foo
1  1 2013-01-02  1.0  3  train  foo
2  1 2013-01-02  1.0  3   test  foo
3  1 2013-01-02  1.0  3  train  foo

In [4]:
df.tail(3)  # prints last 3 entries

A          B    C  D      E    F
1  1 2013-01-02  1.0  3  train  foo
2  1 2013-01-02  1.0  3   test  foo
3  1 2013-01-02  1.0  3  train  foo

- **Information** about our *DataFrame*.

In [5]:
df.dtypes  # prints the data type of each column

A             int64
B    datetime64[ns]
C           float32
D             int32
E          category
F            object
dtype: object

In [6]:
df.index

Int64Index([0, 1, 2, 3], dtype='int64')

In [7]:
df.columns

Index(['A', 'B', 'C', 'D', 'E', 'F'], dtype='object')

In [8]:
df.values

array([[1, Timestamp('2013-01-02 00:00:00'), 1.0, 3, 'test', 'foo'],
       [1, Timestamp('2013-01-02 00:00:00'), 1.0, 3, 'train', 'foo'],
       [1, Timestamp('2013-01-02 00:00:00'), 1.0, 3, 'test', 'foo'],
       [1, Timestamp('2013-01-02 00:00:00'), 1.0, 3, 'train', 'foo']],
      dtype=object)

- **Statistical summary** of our data.

In [9]:
df.describe()  # only numerical features appear when doing this

A    C    D
count  4.0  4.0  4.0
mean   1.0  1.0  3.0
std    0.0  0.0  0.0
min    1.0  1.0  3.0
25%    1.0  1.0  3.0
50%    1.0  1.0  3.0
75%    1.0  1.0  3.0
max    1.0  1.0  3.0

### Sorting and rearanging

- **Transposing** our data.

In [10]:
df.T

0                    1                    2  \
A                    1                    1                    1   
B  2013-01-02 00:00:00  2013-01-02 00:00:00  2013-01-02 00:00:00   
C                    1                    1                    1   
D                    3                    3                    3   
E                 test                train                 test   
F                  foo                  foo                  foo   

                     3  
A                    1  
B  2013-01-02 00:00:00  
C                    1  
D                    3  
E                train  
F                  foo

- **Sorting** by index or value.

In [11]:
df.sort_index(axis=1, ascending=False)  # sorts horizontally

F      E  D    C          B  A
0  foo   test  3  1.0 2013-01-02  1
1  foo  train  3  1.0 2013-01-02  1
2  foo   test  3  1.0 2013-01-02  1
3  foo  train  3  1.0 2013-01-02  1

In [12]:
df.sort_values(by='B')  # sorts DataFrame according to values from column 'B'

A          B    C  D      E    F
0  1 2013-01-02  1.0  3   test  foo
1  1 2013-01-02  1.0  3  train  foo
2  1 2013-01-02  1.0  3   test  foo
3  1 2013-01-02  1.0  3  train  foo

### Slicing and Selection

- Retrieve a **single column** from the *DataFrame*.

In [13]:
df['B']  # returns row with the label 'B'

0   2013-01-02
1   2013-01-02
2   2013-01-02
3   2013-01-02
Name: B, dtype: datetime64[ns]

In [14]:
df.B  # same thing

0   2013-01-02
1   2013-01-02
2   2013-01-02
3   2013-01-02
Name: B, dtype: datetime64[ns]

- **Slicing**.

In [15]:
df[:3]  # returns first three rows 

A          B    C  D      E    F
0  1 2013-01-02  1.0  3   test  foo
1  1 2013-01-02  1.0  3  train  foo
2  1 2013-01-02  1.0  3   test  foo

- Retrieve a **single entry** from the *DataFrame*.
However, if we wanted to get the third entry in our dataframe, we **can't** just write:
```python
df[2] # indexing does NOT work like this in pandas
```
instead, we need to use the built-in *.loc()* function.


In [16]:
df.loc[2]

A                      1
B    2013-01-02 00:00:00
C                      1
D                      3
E                   test
F                    foo
Name: 2, dtype: object

- **Filtering**

In [17]:
df[df['E'] == 'train']  # returns a slice of the DataFrame, where row 'E' contains only the value 'train'

A          B    C  D      E    F
1  1 2013-01-02  1.0  3  train  foo
3  1 2013-01-02  1.0  3  train  foo

### Dealing with missing data.

pandas primarily uses the value `np.nan` to represent missing data.

Reindexing allows you to change/add/delete the index on a specified axis. This returns a copy of the data.

Handling missing data:

- Drop any rows that have missing data: `df.dropna(how='any')`
- Fill missing data: `df.fillna(value=5)`
- Deal with them manually (e.g filling them with the most common element or the mean of the other elements in that column)

### Statistical information

These work only for numerical values. A sample of them are presented below, while there are [many more](https://pandas.pydata.org/pandas-docs/stable/api.html#api-dataframe-stats) available. 

In [18]:
df.sum()  # sum of each column
df.mean()  # mean of each column
df.min()  # minimum element of each column
df.max()  # maximum element of each column
df.std()  # standard deviationn of each column
df.var()  # variance of each column

A    0.0
C    0.0
D    0.0
dtype: float64

We can also apply any function we want (even one that we created) to the dataframe easily:

```python
df.apply(function)
```

### Exercise 1

Fill the missing values of the given DataFrame, with the mean of each value's column.

In [19]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
data = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0,1,2,3])
data[np.random.randint(150, size=20), np.random.randint(4, size=20)] = np.nan
data = pd.DataFrame(data, columns=['A', 'B', 'C', 'D'])

### Solution

In [20]:
data.fillna(data.mean(), inplace=True)  # the 'inplace' argument alows us to avoid 'data = data.fillna(...)' syntax 

### Histograms

In [21]:
s = pd.Series(np.random.randint(0, 7, size=10))
print(s)

s.value_counts()  # creates a histogram of series s

0    3
1    3
2    2
3    3
4    4
5    0
6    3
7    3
8    1
9    4
dtype: int32


3    5
4    2
2    1
1    1
0    1
dtype: int64

### Merging DataFrames

- We can concatenate DataFrames with `pd.concat(list_of_dataframes)`
- We can perform SQL-like joins with `pd.merge(left, right, how)`, where left and right are the two DataFrames we want merged and how refers to the type of the join we want (inner, outer, left, right).

### Grouping data in DataFrames

This procedure follows three steps:
By “group by” we are referring to a process involving one or more of the following steps

- Splitting the data into groups based on some criteria.
- Applying a function to each group independently.
- Combining the results into a data structure.

In [22]:
df = pd.DataFrame({'A' : ['foo', 'bar', 'foo', 'bar',
                          'foo', 'bar', 'foo', 'foo'],
                   'B' : ['one', 'one', 'two', 'three',
                          'two', 'two', 'one', 'three'],
                   'C' : np.random.randn(8),
                   'D' : np.random.randn(8)}) 

print(df)

     A      B         C         D
0  foo    one -0.205790 -0.181443
1  bar    one  0.372470 -0.342720
2  foo    two  1.172693 -0.356442
3  bar  three -1.657100  2.239778
4  foo    two -1.404828  0.624229
5  bar    two  1.043714 -0.277865
6  foo    one -0.591442  1.828205
7  foo  three  0.649791  0.034515


In [23]:
df.groupby(['A','B']).sum()  # Grouping by multiple columns forms a hierarchical index, which we then apply the function.

C         D
A   B                        
bar one    0.372470 -0.342720
    three -1.657100  2.239778
    two    1.043714 -0.277865
foo one   -0.797232  1.646763
    three  0.649791  0.034515
    two   -0.232135  0.267787

### Shape manipulation

In [24]:
# The stack() method “compresses” a level in the DataFrame’s columns.
stk = df.stack()
print(stk)

0  A          foo
   B          one
   C     -0.20579
   D    -0.181443
1  A          bar
   B          one
   C      0.37247
   D     -0.34272
2  A          foo
   B          two
   C      1.17269
   D    -0.356442
3  A          bar
   B        three
   C      -1.6571
   D      2.23978
4  A          foo
   B          two
   C     -1.40483
   D     0.624229
5  A          bar
   B          two
   C      1.04371
   D    -0.277865
6  A          foo
   B          one
   C    -0.591442
   D      1.82821
7  A          foo
   B        three
   C     0.649791
   D    0.0345147
dtype: object


In [25]:
# The inverse operation is unstack()
stk.unstack()

A      B         C          D
0  foo    one  -0.20579  -0.181443
1  bar    one   0.37247   -0.34272
2  foo    two   1.17269  -0.356442
3  bar  three   -1.6571    2.23978
4  foo    two  -1.40483   0.624229
5  bar    two   1.04371  -0.277865
6  foo    one -0.591442    1.82821
7  foo  three  0.649791  0.0345147

### Pivot Tables

In [26]:
df = pd.DataFrame({'A' : ['one', 'one', 'two', 'three'] * 3,
                   'B' : ['A', 'B', 'C'] * 4,
                   'C' : ['foo', 'foo', 'foo', 'bar', 'bar', 'bar'] * 2,
                   'D' : np.random.randn(12),
                   'E' : np.random.randn(12)})

# producing pivot tables is very easy in pandas
pd.pivot_table(df, values='D', index=['A', 'B'], columns=['C'])

C             bar       foo
A     B                    
one   A  0.264018 -0.150525
      B -0.965302  1.470556
      C -0.135368 -1.457308
three A  0.490089       NaN
      B       NaN  2.465016
      C -0.543367       NaN
two   A       NaN -0.710981
      B -0.451513       NaN
      C       NaN  0.374410

### Input operations

- **CSV**.
    Pandas can read csv files and load them to a *DataFrame* through a simple function: `pd.read_csv('file.csv')`. Pandas by default considers the first row in the csv to be the column headers. If there is no such thing in the csv you should use the argument `header=None`. 
- **Excel**.
    Pandas can read and write to excel files through two simple functions: `pd.read_excel(file.xlsx)` and `pd.to_excel(file.xlsx)`. Note that this requires an extra library (xlrd)
- Other options include, SQL databases, clipboard, URLs and even integration with the google analytics API.

## Exploratory Data Analysis

We've only scratched the surface of the capabilities of the pandas library. In order to get a better understanding of the library and how it's used, we'll attempt to perform an exploratory data analysis on the dataset we used for our first exercise.

When doing Exploratory Data Analysis (EDA), we want to observe and summarize our data through descriptive statistics so that we have a better understanding of them. 

In [27]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data'
data = pd.read_csv(url, header=None)
data.columns = ['age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country', 'income']

The first thing we want to do is inspect the shape of the dataframe.

In [28]:
data.shape

(32561, 15)

Our data contains 32561 rows and 15 columns. If we take a look at the [description](https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.names) of the dataset we see that it contains both continuous valued variables (age, working hours etc.) and categorical ones (sex, relationship etc.). When performing data analysis it is important to know what each variable represents.

The next thing we'll do is to look at a sample of the dataset.

In [29]:
data.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

For each variable we'll see what values it can take.

In [30]:
u, c = np.unique(data.age, return_counts=True)
for z in zip(list(u), list(c)):
    print('{}: {}'.format(z[0], z[1]))

17: 395
18: 550
19: 712
20: 753
21: 720
22: 765
23: 877
24: 798
25: 841
26: 785
27: 835
28: 867
29: 813
30: 861
31: 888
32: 828
33: 875
34: 886
35: 876
36: 898
37: 858
38: 827
39: 816
40: 794
41: 808
42: 780
43: 770
44: 724
45: 734
46: 737
47: 708
48: 543
49: 577
50: 602
51: 595
52: 478
53: 464
54: 415
55: 419
56: 366
57: 358
58: 366
59: 355
60: 312
61: 300
62: 258
63: 230
64: 208
65: 178
66: 150
67: 151
68: 120
69: 108
70: 89
71: 72
72: 67
73: 64
74: 51
75: 45
76: 46
77: 29
78: 23
79: 22
80: 22
81: 20
82: 12
83: 6
84: 10
85: 3
86: 1
87: 1
88: 3
90: 43


In [31]:
u, c = np.unique(data.workclass, return_counts=True)
for z in zip(list(u), list(c)):
    print('{}: {}'.format(z[0], z[1]))

 ?: 1836
 Federal-gov: 960
 Local-gov: 2093
 Never-worked: 7
 Private: 22696
 Self-emp-inc: 1116
 Self-emp-not-inc: 2541
 State-gov: 1298
 Without-pay: 14


Here we find our first occurrence  of missing values. In this dataset, these are represented by question marks (?).

In [32]:
len(set(data.fnlwgt))

21648

This variable is continuous-valued and represents the demographics of the individual. 

>Description of fnlwgt (final weight)

> The weights on the CPS files are controlled to independent estimates of the
  civilian noninstitutional population of the US.  These are prepared monthly
  for us by Population Division here at the Census Bureau.  We use 3 sets of
  controls.
   These are:
           1.  A single cell estimate of the population 16+ for each state.
           2.  Controls for Hispanic Origin by age and sex.
           3.  Controls by Race, age and sex.
 
> We use all three sets of controls in our weighting program and "rake" through
  them 6 times so that by the end we come back to all the controls we used.
 
> The term estimate refers to population totals derived from CPS by creating
  "weighted tallies" of any specified socio-economic characteristics of the
  population.
 
> People with similar demographic characteristics should have
  similar weights.  There is one important caveat to remember
  about this statement.  That is that since the CPS sample is
  actually a collection of 51 state samples, each with its own
  probability of selection, the statement only applies within
  state.

In [33]:
u, c = np.unique(data.education, return_counts=True)
for z in zip(list(u), list(c)):
    print('{}: {}'.format(z[0], z[1]))

 10th: 933
 11th: 1175
 12th: 433
 1st-4th: 168
 5th-6th: 333
 7th-8th: 646
 9th: 514
 Assoc-acdm: 1067
 Assoc-voc: 1382
 Bachelors: 5355
 Doctorate: 413
 HS-grad: 10501
 Masters: 1723
 Preschool: 51
 Prof-school: 576
 Some-college: 7291


In [34]:
u, c = np.unique(data['education-num'], return_counts=True)
for z in zip(list(u), list(c)):
    print('{}: {}'.format(z[0], z[1]))

1: 51
2: 168
3: 333
4: 646
5: 514
6: 933
7: 1175
8: 433
9: 10501
10: 7291
11: 1382
12: 1067
13: 5355
14: 1723
15: 576
16: 413


In [35]:
u, c = np.unique(data['marital-status'], return_counts=True)
for z in zip(list(u), list(c)):
    print('{}: {}'.format(z[0], z[1]))

 Divorced: 4443
 Married-AF-spouse: 23
 Married-civ-spouse: 14976
 Married-spouse-absent: 418
 Never-married: 10683
 Separated: 1025
 Widowed: 993


In [36]:
u, c = np.unique(data.occupation, return_counts=True)
for z in zip(list(u), list(c)):
    print('{}: {}'.format(z[0], z[1]))

 ?: 1843
 Adm-clerical: 3770
 Armed-Forces: 9
 Craft-repair: 4099
 Exec-managerial: 4066
 Farming-fishing: 994
 Handlers-cleaners: 1370
 Machine-op-inspct: 2002
 Other-service: 3295
 Priv-house-serv: 149
 Prof-specialty: 4140
 Protective-serv: 649
 Sales: 3650
 Tech-support: 928
 Transport-moving: 1597


In [37]:
u, c = np.unique(data.relationship, return_counts=True)
for z in zip(list(u), list(c)):
    print('{}: {}'.format(z[0], z[1]))

 Husband: 13193
 Not-in-family: 8305
 Other-relative: 981
 Own-child: 5068
 Unmarried: 3446
 Wife: 1568


In [38]:
u, c = np.unique(data.race, return_counts=True)
for z in zip(list(u), list(c)):
    print('{}: {}'.format(z[0], z[1]))

 Amer-Indian-Eskimo: 311
 Asian-Pac-Islander: 1039
 Black: 3124
 Other: 271
 White: 27816


In [39]:
u, q = np.unique(data.sex, return_counts=True)
for z in zip(list(u), list(q)):
    print('{}: {}'.format(z[0], z[1]))

 Female: 10771
 Male: 21790


In [40]:
print(len(data[data['capital-gain'] == 0]))
print(len(data[data['capital-gain'] != 0]))

29849
2712


In [41]:
print(len(data[data['capital-loss'] == 0]))
print(len(data[data['capital-loss'] != 0]))

31042
1519


In [42]:
len(set(data['hours-per-week']))

94

In [43]:
u, c = np.unique(data['native-country'], return_counts=True)
for z in zip(list(u), list(c)):
    print('{}: {}'.format(z[0], z[1]))

 ?: 583
 Cambodia: 19
 Canada: 121
 China: 75
 Columbia: 59
 Cuba: 95
 Dominican-Republic: 70
 Ecuador: 28
 El-Salvador: 106
 England: 90
 France: 29
 Germany: 137
 Greece: 29
 Guatemala: 64
 Haiti: 44
 Holand-Netherlands: 1
 Honduras: 13
 Hong: 20
 Hungary: 13
 India: 100
 Iran: 43
 Ireland: 24
 Italy: 73
 Jamaica: 81
 Japan: 62
 Laos: 18
 Mexico: 643
 Nicaragua: 34
 Outlying-US(Guam-USVI-etc): 14
 Peru: 31
 Philippines: 198
 Poland: 60
 Portugal: 37
 Puerto-Rico: 114
 Scotland: 12
 South: 80
 Taiwan: 51
 Thailand: 18
 Trinadad&Tobago: 19
 United-States: 29170
 Vietnam: 67
 Yugoslavia: 16


## Data Preparation

Next, well look at several ways we might have to manipulate our data, including data cleaning, imputing and transforming 

Because the unknown values are represented as question marks this dataset, we need to handle them. The most common ways of dealing with missing data are:

- Removing the whole row.
- Filling them with the median or mean value (for continuous variables).
- Filling them with the most common value (for categorical variables).
- Treat them as a new category.
- Running a ML algorithm to predict those values (we'll examine this in future lessons).

### Example: fill the occupation with the most frequent element

In [44]:
u, c = np.unique(data.occupation, return_counts=True)
most_freq = u[np.argmax(q)]  # find the most common element
data.occupation = data.occupation.apply(lambda x: most_freq if x == ' ?' else x)
# the line above first keeps just the column that represents the occupations from the dataframe
# then it applies a function which checks if those values are question marks and changes them to the most common element
# finally it replaces the original occupations with the new ones
u, c = np.unique(data.occupation, return_counts=True)
for z in zip(list(u), list(c)):
    print('{}: {}'.format(z[0], z[1]))

 Adm-clerical: 5613
 Armed-Forces: 9
 Craft-repair: 4099
 Exec-managerial: 4066
 Farming-fishing: 994
 Handlers-cleaners: 1370
 Machine-op-inspct: 2002
 Other-service: 3295
 Priv-house-serv: 149
 Prof-specialty: 4140
 Protective-serv: 649
 Sales: 3650
 Tech-support: 928
 Transport-moving: 1597


`.apply()` is not limited to filling missing values. It can also assist in transforming the data. For instance we can use it to strip the whitespace in the occupations: 

In [45]:
print('Before cleaning: `{}`'.format(data.occupation[0]))
data.occupation = data.occupation.apply(lambda x: x.strip())
print('After cleaning: `{}`'.format(data.occupation[0]))

Before cleaning: ` Adm-clerical`
After cleaning: `Adm-clerical`


### Exercise 2

Fill the missing values of the DataFrame's `native-country` column with whatever strategy you wish.

### Solution

This time we'll drop the rows containing missing values.

In [46]:
print('dataframe length:', len(data))
print('missing:', len(data[data['native-country'] == ' ?']))
data.drop(data.index[data['native-country'] == ' ?'], inplace=True)
print('dataframe length:', len(data))
print('missing:', len(data[data['native-country'] == ' ?']))

dataframe length: 32561
missing: 583
dataframe length: 31978
missing: 0


Finally, we'll look at how we can prepare our variables for machine learning. These models require numeric vectors as input. This means that all text data needs to be encoded as numbers.

**Categorical data**:

 - When the data doesn't have a sequential nature (e.g. country of origin) we need to one-hot encode our variables. This is done to ensure that the model doesn't try to exploit sequential traits when there are none available. An example of this is depicted in the image below:
 
 ![](https://i.imgur.com/mtimFxh.png)
 
 For this to be done, we need to create a new variable for each unique value, thus increasing the number of variables in the dataset. This may have some complications, we will discuss in a future lesson. 
 
 
 - When the data does have a sequential nature (e.g. ratings: poor/average/good), we can just map our unique values to integers, while preserving their nature.
 
 
 - We may also elect to perform a custom encoding on our data specific to the problem.
 
**Numerical data**:

 - Usually we can leave them as is.
 
 
 - If we aren't pleased with their distribution we can convert them to categorical by binning (or bucketing) them.
 
 
 
To illustrate the above and see how they would be performed  in pandas: We will first encode the `education` variable preserving its sequential nature. Next, we will perform a custom encoding on the `marital-status` variable so that we keep only two categories (i.e. currently married and not). Finally, we will one-hot encode all the remaining categorical variables in the dataset. 

In [47]:
data.education = data.education.apply(lambda x: x.strip())  # clean whitespace on category


### Encode education variable ###

# Create a dictionary mapping the categories to their encodings.
# This has to be done manually as the exact sequence has to be taken into consideration.
mappings = {'Preschool': 1, '1st-4th': 2, '5th-6th': 3, '7th-8th': 4, '9th': 5, '10th': 6,
            '11th': 7, '12th': 8, 'HS-grad': 9, 'Some-college': 10, 'Assoc-voc': 11, 'Assoc-acdm':12,
             'Bachelors': 13, 'Masters': 14, 'Prof-school': 15, 'Doctorate': 16}

data.replace(mappings, inplace=True)  # encode categorical variable with custom mapping
# another way this could be done would be through data.education.astype('category'),
# this however would prevent us from choosing the mapping scheme


### Encode marital-status variable ###

data['marital-status'] = np.where(data["marital-status"] == ' Married-civ-spouse', 1, 0)
# the above function replaces ' Married-civ-spouse' with 1 and all the rest with 0


### Encode rest variables ###

data = pd.get_dummies(data)  # one-hot encode all categorical variables

Finally, we'll see how we can split numerical values to separate bins, in order to convert them to categorical. This time around we won't replace the numerical data but create a new variable.

In [48]:
data['age_categories'] = pd.cut(data.age, 3, labels=['young', 'middle aged', 'old'])
pd.value_counts(data['age_categories'])

young          19556
middle aged    11280
old             1142
Name: age_categories, dtype: int64

When binning would usually want each bin to have the same number of samples. In order to do this we need to manually find there to cut each bin input the *cut points* instead of the number of bins we want. But, we'll leave that up to you!

Finally, let's save the DataFrame for future use.

### Additional material:

- Advanced filtering. We can use `.apply()` to create more complex filters. Say we want to filter the DataFrame by people whose names start with the letter *A*.

```python
# Basic syntax:
def starts_with_a(word):
    if word[0] == 'A':
        return True
    else:
        return False
    
df[df['names'].apply(starts_with_a)]

# Using lambda functions:
df[df['names'].apply(lambda x: x[0] == 'A')]
```

- Data wrangling:
[extended data wrangling tutorial](http://nbviewer.jupyter.org/github/fonnesbeck/Bios8366/blob/master/notebooks/Section2_2-Data-Wrangling-with-Pandas.ipynb)

## Dealing with inconsistent text data

One of the most common problems when dealing with text data is inconsistency. This may occur due to spelling errors, differences when multiple people perform the data entry, etc.

In [49]:
df = pd.DataFrame({'fname':['George', 'george ', 'GEORGIOS', 'Giorgos', ' Peter', 'Petet'],
                   'sname':['Papadopoulos', 'alexakos ', 'Georgiou', 'ANTONOPOULOS', ' Anastasiou', 'Κ'],
                   'age': [46, 34, 75, 24, 54, 33]})
df

age     fname         sname
0   46    George  Papadopoulos
1   34   george      alexakos 
2   75  GEORGIOS      Georgiou
3   24   Giorgos  ANTONOPOULOS
4   54     Peter    Anastasiou
5   33     Petet             Κ

When looking at the example above, several inconsistencies become apparent. The first thing we want to do when dealing with strings is to convert them all to lowercase (or uppercase depending on preference) and remove preceding and succeeding whitespace.

In [50]:
def clean_text(text):
    text = text.strip()  # strip whitespace
    text = text.lower()  # convert to lowercase
    return text

df.fname = df.fname.apply(clean_text)
df.sname = df.sname.apply(clean_text)

# same could be done through a lambda function
# df.fname.apply(lambda x: x.strip().lower())

df

age     fname         sname
0   46    george  papadopoulos
1   34    george      alexakos
2   75  georgios      georgiou
3   24   giorgos  antonopoulos
4   54     peter    anastasiou
5   33     petet             κ

Another problem originates from the way each name was entered. There are two ways to deal with this, one is to manually look and change errors, and the other is to compare the strings to find differences.

We are going to try the second, through the python package [fuzzywuzzy](https://github.com/seatgeek/fuzzywuzzy).

In [51]:
from fuzzywuzzy import process
process.extract('george', df.fname, limit=None)

c:\users\thano\appdata\local\programs\python\python36\lib\site-packages\fuzzywuzzy\fuzz.py:35: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')


[('george', 100, 0),
 ('george', 100, 1),
 ('georgios', 71, 2),
 ('giorgos', 62, 3),
 ('peter', 36, 4),
 ('petet', 36, 5)]

Fuzzywuzzy compares strings and outputs a score depending on how close they are. Let's replace the close ones:

In [52]:
def replace_matches_in_column(df, column, target_string, min_ratio=50):
    
    # find unique elements in specified column
    strings = df[column].unique()
    
    # see how close these elements are to the target string
    matches = process.extract(target_string, strings, limit=None)

    # keep only the closest ones
    close_matches = [matches[0] for matches in matches if matches[1] >= min_ratio]

    # get the rows of all the close matches in our dataframe
    rows_with_matches = df[column].isin(close_matches)

    # replace all rows with close matches with the input matches 
    df.loc[rows_with_matches, column] = target_string

In [53]:
replace_matches_in_column(df, 'fname', 'george')
replace_matches_in_column(df, 'fname', 'peter')
df

age   fname         sname
0   46  george  papadopoulos
1   34  george      alexakos
2   75  george      georgiou
3   24  george  antonopoulos
4   54   peter    anastasiou
5   33   peter             κ

There, all clean! Note that the `min_ratio` we used is **very low**. We usually require much closer matches in order to replace.